### PR TITLE
Fix/query endpoint timeout

### LINKED ISSUE #1033: Ohio practitioner not visible to Ohio state staff
### Notes

@amballon  discussed with @landonshumway-ia and needs further investigation.

1. Log in as Abril Bell, an OT/OTA practitioner based in Ohio, with andrea.tringo+jcc3@inspiringapps.com / 1045Pearl!$!$ to confirm home state (Ohio), status (Active), etc.
2. Log in as Ohio state staff, andrea.tringo+devstate@inspiringapps.com (same PW) and search for all licensing data in Ohio
3. Log in as Compact staff, andrea.tringo@inspiringapps.com, select Occupational Therapy from bottom left nav and search for all licensing data in Ohio

Expected: Ohio state staff sees Abril Bell within the list of user as does the Compact staff
Actual: Abril Bell is not in the state staff list of users/pratitioners

**_Abril Bell's license status and verification letter:_**

Image

[ABell_verification.pdf](https://github.com/user-attachments/files/21972471/ABell_verification.pdf)

**_State staff's view of all Ohio practitioners:_**

Image

**_Compact staff's view of all Ohio practitioners (OT)_**

Image

### Questions
1. **Q: Short question text...**
    - _More info & examples..._
    - **A:** ???

### Assumptions
1. Assumption text...

### Estimate: `?? hrs`
- Item 1: `?? hrs`
- TODO

### Tasks
- [ ] TODO
---
- [ ] Cross browser device testing
- [ ] Confirm `yarn build` is error & warning free
- [ ] Confirm `yarn serve` is error & warning free
- [ ] Confirm `yarn test:unit:all` is error & warning free
- [ ] Confirm `yarn test:e2e` is error & warning free
- [ ] QA Notes added
- [ ] PR opened with labels / reviewers / assignee / linked-issue

### QA Notes
- TODO

### PR DESCRIPTION
**Pull request by landonshumway-ia**
We have identified some issues with our pagination logic which needed to be addressed. 

The first issue was our truncation logic for returning a specific number of items was dropping matching records from the last page of results by not returning a pagination key, so the client was unable to get the final few records for a result set. 

The next issue was in regards to how we were using the limit parameter for our DynamoDB queries. The assumption was made that the limit parameter determined how many records to return, but the actual purpose of the limit parameter is to set how many records to evaluate for matches when making a query. For queries that don't use filter expressions, this is not a problem since each item evaluated will always match. However, for cases like searching for providers just by state, we heavily rely on filter expressions which means that this pagination logic was making thousands of requests and timing out since it was using a limit of 25 records per query.

Long term, this search functionality is intended to be replaced by a more robust DB solution designed for complex queries (ie OpenSearch) Until then, this updates our pagination decorator logic to handle large data sets with filter expressions so the search endpoint can find enough matches for a page before timing out.

Closes #1033 

## Summary by CodeRabbit

- Bug Fixes
  - Pagination behavior improved: paginator better adheres to requested page sizes and returns correct next-page keys when filters reduce results.

- New Features
  - Public provider search now returns an empty providers list (preserving query/sorting/pagination) when a jurisdiction isn’t marked live.
  - Added a utility to determine whether a jurisdiction is live in a compact.

- Tests
  - Added and updated pagination tests covering client/server-filter edge cases; removed a legacy duplicated suite.

- Chores
  - Removed several deployment-time migration scripts and related wiring; granted compact-configuration read access to public lookup.

### CODE REVIEW

File: backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py [hunk 1]
```diff
@@ -22,14 +22,29 @@ class paginated_query:  # noqa: N801 invalid-name
...
-    When a FilterExpression is used, DynamoDB can return fewer items than is specified as the pageSize. To ensure
-    that we always return the full pageSize, when there are enough results, this decorator will also potentially repeat
-    queries multiple times internally, until it has pageSize items to return.
+    IMPORTANT: When a FilterExpression is used over a large partition space, DynamoDB can return fewer items than is
+    specified as the pageSize. To ensure that we always return the full pageSize, this decorator will repeat queries as
+    needed until it has a full page of items to return. In order to reduce the number of queries made when using filter
+    expressions, you should set the set_query_limit_to_match_page_size flag to False. This will not set the Limit
+    parameter on the query, so DynamoDB will evaluate as many items as it can within a single query, returning all
+    evaluated items that match the filter expression. The decorator will handle truncating the items to fit within the
+    pageSize, and will also handle calculating the lastKey for the next page of results.
```
[the remark this answers is not in the document]
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py:35]: updated

File: backend/compact-connect/lambdas/python/provider-data-v1/handlers/public_lookup.py [hunk 2]
```diff
@@ -45,6 +45,23 @@ def public_query_providers(event: dict, context: LambdaContext):  # noqa: ARG001
 
         jurisdiction = query.get('jurisdiction')
 
+        if jurisdiction:
```
Comment (jusdino) [inline backend/compact-connect/lambdas/python/provider-data-v1/handlers/public_lookup.py:48]: Good thought.

**Review by jusdino — APPROVED** [review]
Yep, that'll fix it. Did we get a follow-up ticket in the backlog for a longer-term solution after we have flexible search built?

File: backend/compact-connect/lambdas/python/common/tests/function/test_data_model/test_paginated.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+from .. import TstFunction
+
+MOCK_PROVIDER_ID_PREFIX = '89a6377e-c3a5-40e5-bca5-317ec854c5'
+
+
+@mock_aws
+class TestPaginated(TstFunction):
+    def test_pagination_returns_pagination_key_if_more_items_than_page_size_on_first_query(self):
+        """Test edge case: page size 5, 10 providers total, no limit set, 8 providers match filter.
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/common/tests/function/test_data_model/test_paginated.py:15]: What does "no limit set" mean, and where does that pop up in the code below?

File: backend/compact-connect/lambdas/python/common/tests/function/test_data_model/test_paginated.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        # now we call again with the last key and ensure we get the remaining 3 providers
+        resp = self.config.data_client.get_providers_sorted_by_family_name(
+            compact='aslp', pagination={'pageSize': 5, 'lastKey': resp['pagination']['lastKey']}, jurisdiction='ky'
+        )
+        self.assertEqual(len(resp['items']), 3)
+        self.assertIsNone(resp['pagination']['lastKey'])
+
+    def test_pagination_does_not_return_pagination_key_if_number_of_items_is_exact_match_to_page_size(self):
+        """Test edge case: page size 5, 5 providers total, flag set to False, all 5 providers match filter.
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/common/tests/function/test_data_model/test_paginated.py:65]: What does "Flag set to False" mean, and where does that pop up in the code below?

File: backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py [hunk 5]
```diff
@@ -56,23 +76,38 @@ def __call__(self, *args, pagination: dict = None, client_filter: Callable[[dict
...
+            items = items[:page_size]
+            last_item = items[-1]
+            # Since we truncated our items, we need to recalculate the last key
+            # We will use the last known evaluated key to determine the needed fields
+            # for the last key if we have one
+            if last_known_evaluated_key is not None:
+                last_key = {k: last_item[k] for k in last_known_evaluated_key.keys()}
+            # else the first query was the last query, but there were more items returned than the page size
+            # in this case, we need to determine the last key by making a query with a limit of 1
+            # so we can get the last key from the response and use it for the last item in the list
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py:93]: It seems we are only doing this to get the field names of the `LastEvaluatedKey` field – aren't those static fields that we already know?

**Review by jlkravitz — COMMENTED** [review]
A few questions/nits!

File: backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py (hunk 5 above)
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py:93]: No, in the case of querying by index, you must include the index fields as part of the LastEvalutatedKey. I will add more detail to the comment to hopefully make this clear.

File: backend/compact-connect/lambdas/python/common/tests/function/test_data_model/test_paginated.py (hunk 4 above)
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/python/common/tests/function/test_data_model/test_paginated.py]: I've reworded this to hopefully make it more clear

File: backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py [hunk 6]
```diff
@@ -56,23 +76,41 @@ def __call__(self, *args, pagination: dict = None, client_filter: Callable[[dict
...
+            # size. In this case, we need to determine the last key for getting the next page. The keys needed by the
+            # last key are not static (for example, if you are querying by a GSI you must include the GSI fields as
+            # part of the last key) so to have a generic solution for this scenario, we make a query with a limit of 1
+            # so we can get the LastEvaluatedKey from the response, and then map the keys from that to the values of
+            # the last item that will be returned in the response to create our own last key.
+            else:
+                last_key_resp = self._caught_query(client_filter, *args, dynamo_pagination={'Limit': 1}, **kwargs)
+                last_known_evaluated_key = last_key_resp.get('LastEvaluatedKey')
+                last_key = {k: last_item[k] for k in last_known_evaluated_key.keys()}
+        # else if the page size matched the query, and there are more records to return, set the last key to match
```
[the remark this answers is not in the document]
Comment (jlkravitz) [reply backend/compact-connect/lambdas/python/common/cc_common/data_model/query_paginator.py:101]: @landonshumway-ia PR looks good to me but flagging this comment. Feel free to resolve if you decide it's not worth implementing or a false positive, etc.

**Review by jlkravitz — APPROVED** [review]
@isabeleliassen good to merge!
